### PR TITLE
feat(rdpdr): simple file backend for file upload/download

### PR DESCRIPTION
**Pull request by zmrush**
As mentiond in [file upload support](https://github.com/Devolutions/IronRDP/issues/110). i implemented a simple file backend,which supports file upload/download. @CBenoit @ibeckermayer

### CODE REVIEW

**Review by CBenoit — COMMENTED** [review]
Thank you for moving this forward!!

Could you please put the backend implementation in a separate crate named `ironrdp-rdpdr-native`?

See for instance [`ironrdp-cliprdr-native`](https://github.com/Devolutions/IronRDP/tree/abc0ee0617fcdcda3f0ace3ffd292251979d1910/crates/ironrdp-cliprdr-native)

Also, try to run `cargo xtask ci` to run the various checks we are enforcing in our CI.

**Comment by zmrush** [comment]
> Thank you for moving this forward!!
> 
> Could you please put the backend implementation in a separate crate named `ironrdp-rdpdr-native`?
> 
> See for instance [`ironrdp-cliprdr-native`](https://github.com/Devolutions/IronRDP/tree/abc0ee0617fcdcda3f0ace3ffd292251979d1910/crates/ironrdp-cliprdr-native)
> 
> Also, try `cargo xtask ci` to run the various checks we are enforcing in our CI.

ok,i will try it. thank u

File: crates/ironrdp-rdpdr-native/README.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# IronRDP RDPDR native backends
+
+Native CLIPRDR backend implementations. Currently only *nix systems are supported.
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/README.md:3]:
```suggestion
Native RDPDR backend implementations. Currently only *nix systems are supported.
```

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 2]
```diff
@@ -0,0 +1,765 @@
...
+    pub file_base: String,
+    pub file_map: std::collections::HashMap<u32, std::fs::File>,
+    pub file_path_map: std::collections::HashMap<u32, String>,
+    pub file_dir_map: std::collections::HashMap<u32, DirBox>,
+}
+#[derive(Debug)]
+pub struct DirBox {
+    pub dir: *mut DIR,
+}
+unsafe impl Send for DirBox {}
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:28]: issue: This need to be documented with an `SAFETY` comment.

File: crates/ironrdp-rdpdr-native/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,4 @@
+#[cfg(any(target_os = "macos", target_os = "linux"))]
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/lib.rs:1]: suggestion: Can you add this at the top of the file and fix the new warnings?

```rust
#![doc = include_str!("../README.md")]
#![warn(unsafe_op_in_unsafe_fn)]
#![warn(invalid_reference_casting)]
#![warn(pointer_structural_match)]
#![warn(clippy::undocumented_unsafe_blocks)]
#![warn(clippy::multiple_unsafe_ops_per_block)]
#![warn(clippy::transmute_ptr_to_ptr)]
#![warn(clippy::as_ptr_cast_mut)]
#![warn(clippy::cast_ptr_alignment)]
#![warn(clippy::fn_to_numeric_cast_any)]
#![warn(clippy::ptr_cast_constness)]
```

File: crates/ironrdp-rdpdr-native/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+test = false
+
+[dependencies]
+ironrdp-error.workspace = true
+ironrdp-pdu.workspace = true
+ironrdp-svc.workspace = true
+ironrdp-rdpdr.workspace = true
+tracing.workspace = true
+bitflags.workspace = true
+nix = { version = "0.29" ,features = ["fs"]}
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/Cargo.toml:25]: nitpick:
```suggestion
nix = { version = "0.29", features = ["fs"] }
```

File: crates/ironrdp-rdpdr-native/src/nix/mod.rs [hunk 5]
```diff
@@ -0,0 +1,2 @@
+#[allow(clippy::unnecessary_cast)]
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/mod.rs:1]: question: Why did you allow this warning? Could you add a comment explaining why directly in the code?

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 6]
```diff
@@ -0,0 +1,765 @@
...
+
+use ironrdp_rdpdr::pdu::efs::*;
+use ironrdp_rdpdr::pdu::esc::{ScardCall, ScardIoCtlCode};
+use ironrdp_rdpdr::pdu::RdpdrPdu;
+use ironrdp_rdpdr::RdpdrBackend;
+use std::ffi::{CStr, CString};
+use std::io::{Seek, SeekFrom, Write};
+use std::os::fd::{AsFd, AsRawFd};
+use std::os::unix::fs::MetadataExt;
+use tracing::{debug, warn};
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:14]: suggestion: Put this in lib.rs so we don’t need to `use` the logging macros. Usually I don’t recommend to do that expect for tracing.

```rust
#[macro_use]
extern crate tracing;
```

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 7]
```diff
@@ -0,0 +1,765 @@
...
+                    warn!("write len:{} is not equal to {}", length, req_inner.write_data.len());
+                    let res = RdpdrPdu::DeviceWriteResponse(DeviceWriteResponse {
+                        device_io_reply: DeviceIoResponse::new(request, NtStatus::UNSUCCESSFUL),
+                        length: 0_u32,
+                    });
+                    Ok(vec![SvcMessage::from(res)])
+                }
+            }
+            Err(e) => {
+                warn!("write error:{:?}", e);
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:110]: style: See [this section](https://github.com/Devolutions/IronRDP/blob/abc0ee0617fcdcda3f0ace3ffd292251979d1910/STYLE.md#logging)
```suggestion
            Err(error) => {
                warn!(%error, "Write error");
```

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 8]
```diff
@@ -0,0 +1,765 @@
...
+}
+pub(crate) fn read_device(
+    backend: &mut SimpleRdpdrBackend,
+    req_inner: DeviceReadRequest,
+) -> PduResult<Vec<SvcMessage>> {
+    let read_inner = |file: &mut std::fs::File, offset, length| -> std::io::Result<Vec<u8>> {
+        let sf = SeekFrom::Start(offset);
+        file.seek(sf)?;
+        let mut buf = vec![0; length];
+        use std::io::Read;
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:128]: suggestion: I’m all for putting local `use` inside functions, but this should be at the top of the outermost function for clarity.

```rust
fn read_device() {
    use std::io::Read as _;

    fn helper() {
        // Use std::io::Read
    }
}
```

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 9]
```diff
@@ -0,0 +1,765 @@
...
+    backend: &mut SimpleRdpdrBackend,
+    req_inner: DeviceWriteRequest,
+) -> PduResult<Vec<SvcMessage>> {
+    let write_inner = |file: &mut std::fs::File, offset, write_data| -> std::io::Result<usize> {
+        let sf = SeekFrom::Start(offset);
+        file.seek(sf)?;
+        let length = file.write(write_data)?;
+        file.flush()?;
+        Ok(length)
+    };
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:80]: style: If capturing environment is not required, [this is the style we follow](https://github.com/Devolutions/IronRDP/blob/abc0ee0617fcdcda3f0ace3ffd292251979d1910/STYLE.md#local-helper-functions)

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 10]
```diff
@@ -0,0 +1,765 @@
...
+                    let io_response = DeviceIoResponse::new(req_inner.device_io_request, NtStatus::UNSUCCESSFUL);
+                    let res = RdpdrPdu::DeviceCreateResponse(DeviceCreateResponse {
+                        device_io_reply: io_response,
+                        file_id,
+                        information: Information::empty(),
+                    });
+                    return Ok(vec![SvcMessage::from(res)]);
+                }
+                //这个我们先不执行，我们看看能不能统一
+                //我们可以统一到文件打开的条件中去
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:667]: issue (non-blocking): Use English in the comments, because very few of IronRDP contributor can read Chinese 😆

Something like this? I’ll leave you adjust accordingly.
```suggestion
                // Let's not execute this for now, let's see if we can unify it.
                // This can be unified with the condition for opening the file.
```

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 11]
```diff
@@ -0,0 +1,765 @@
...
+                            backend.file_path_map.insert(file_id, path.clone());
+                            return make_create_drive_resp(
+                                req_inner.device_io_request,
+                                req_inner.create_disposition,
+                                file_id,
+                            );
+                        }
+                        Err(e) => {
+                            warn!("open file dir error:{:?}", e);
+                            //return by downside
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:700]: style: See [this](https://github.com/Devolutions/IronRDP/blob/abc0ee0617fcdcda3f0ace3ffd292251979d1910/STYLE.md#inline-code-comments-are-proper-sentences)

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 12]
```diff
@@ -0,0 +1,765 @@
...
+            });
+            Ok(vec![SvcMessage::from(res)])
+        },
+        |file, request| match write_inner(file, req_inner.offset, &req_inner.write_data) {
+            Ok(length) => {
+                if length == req_inner.write_data.len() {
+                    Ok(vec![SvcMessage::from(RdpdrPdu::DeviceWriteResponse(
+                        DeviceWriteResponse {
+                            device_io_reply: DeviceIoResponse::new(request, NtStatus::SUCCESS),
+                            length: req_inner.write_data.len() as u32,
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:97]: suggestion: `as` conversions should be avoided when possible. We prefer using `u32::from` or `u32::try_from`. It’s fine to use `u32::try_from(…).unwrap()` in most cases, to at least crash when the conversion is unexpectedly lossy.

rationale: `as` conversions will perform many kinds of conversions, including silently lossy conversions and dangerous coercions.

File: crates/ironrdp-rdpdr/src/backend/mod.rs [hunk 13]
```diff
@@ -7,10 +7,11 @@ use ironrdp_svc::AsAny;
...
 use crate::pdu::efs::{DeviceControlRequest, ServerDeviceAnnounceResponse, ServerDriveIoRequest};
 use crate::pdu::esc::{ScardCall, ScardIoCtlCode};
+use ironrdp_svc::SvcMessage;
 
 /// OS-specific device redirection backend interface.
 pub trait RdpdrBackend: AsAny + fmt::Debug + Send {
     fn handle_server_device_announce_response(&mut self, pdu: ServerDeviceAnnounceResponse) -> PduResult<()>;
     fn handle_scard_call(&mut self, req: DeviceControlRequest<ScardIoCtlCode>, call: ScardCall) -> PduResult<()>;
-    fn handle_drive_io_request(&mut self, req: ServerDriveIoRequest) -> PduResult<()>;
+    fn handle_drive_io_request(&mut self, req: ServerDriveIoRequest) -> PduResult<Vec<SvcMessage>>;
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr/src/backend/mod.rs:16]: @ibeckermayer This is slightly changing the API for `RdpdrBackend`. Does that look good to you?

**Review by CBenoit — COMMENTED** [review]
Great work!! Thank you a lot

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 14]
```diff
@@ -0,0 +1,765 @@
...
+    match backend.file_path_map.get(&req_inner.device_io_request.file_id) {
+        Some(parent_pos_for_next) => {
+            let mut find_file_name = None;
+            if req_inner.initial_query > 0 {
+                if req_inner.path.ends_with('*') {
+                    unsafe {
+                        let mut parent = backend.file_base.clone();
+                        let query_path = req_inner.path.replace('\\', "/");
+                        let len = query_path.len();
+                        parent.push_str(&query_path[0..len - 1]);
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:538]: suggestion: FYI, we are using the `clippy::arithmetic_side_effects` to warn us about potential arithmetic side-effect bugs. Typically, we’ll `allow` it and explain why there is no problem with a comment. Alternatively, it’s possible to use checked arithmetic operations.

For instance:

https://github.com/Devolutions/IronRDP/blob/abc0ee0617fcdcda3f0ace3ffd292251979d1910/crates/ironrdp-cliprdr-format/src/bitmap.rs#L497-L501

https://github.com/Devolutions/IronRDP/blob/abc0ee0617fcdcda3f0ace3ffd292251979d1910/crates/ironrdp-web/src/session.rs#L548-L550

See also: https://github.com/Devolutions/IronRDP/blob/abc0ee0617fcdcda3f0ace3ffd292251979d1910/STYLE.md#explain-non-obvious-assumptions-by-referencing-the-invariants

**Comment by CBenoit** [comment]
Modified the title of the PR to follow the conventional commits specification: https://www.conventionalcommits.org/en/v1.0.0/

Also, please push new commits when applying review suggestions. It’s easier for me to review incrementally only what is changed. Of course, it’s fine to clean the history using rebase before or after the review.

File: crates/ironrdp-rdpdr-native/src/lib.rs (hunk 3 above)
Comment (zmrush) [reply crates/ironrdp-rdpdr-native/src/lib.rs:16]: ok, i add these contents

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs (hunk 11 above)
Comment (zmrush) [reply crates/ironrdp-rdpdr-native/src/nix/simple.rs:722]: ok, i modify according to our style

File: crates/ironrdp-rdpdr-native/src/nix/mod.rs (hunk 5 above)
Comment (zmrush) [reply crates/ironrdp-rdpdr-native/src/nix/mod.rs]: because statvfs.blocks_available() return u32/u64 on different platform(for example, on my computer its u32), but volume_serial_number is u32, so i need to use ('as u32' or u32::try_from()),but it emit warning 'unnecessary_cast', so i add this

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs (hunk 6 above)
Comment (zmrush) [reply crates/ironrdp-rdpdr-native/src/nix/simple.rs]: ok,this seems good. i modify the code

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs (hunk 10 above)
Comment (zmrush) [reply crates/ironrdp-rdpdr-native/src/nix/simple.rs]: ok, i am sorry for forgetting to modify this comments

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs (hunk 14 above)
Comment (zmrush) [reply crates/ironrdp-rdpdr-native/src/nix/simple.rs]: ok, it seems these warning will emit after i add clippy warns. And therefore i modify the code

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs (hunk 7 above)
Comment (zmrush) [reply crates/ironrdp-rdpdr-native/src/nix/simple.rs]: ok, i modify according to our style

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs (hunk 9 above)
Comment (zmrush) [reply crates/ironrdp-rdpdr-native/src/nix/simple.rs]: ok, i modify according to our style

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs (hunk 12 above)
Comment (zmrush) [reply crates/ironrdp-rdpdr-native/src/nix/simple.rs]: yes,it seems reasonable. And i modify it

**Comment by zmrush** [comment]
> Modified the title of the PR to follow the conventional commits specification: https://www.conventionalcommits.org/en/v1.0.0/
> 
> Also, please push new commits when applying review suggestions. It’s easier for me to review incrementally only what is changed. Of course, it’s fine to clean the history using rebase before or after the review.

ok,i modify the the title of the PR. but i do not know how to 'Also, please push new commits when applying review suggestions'. i only reply u on this 'Conversation' page and 'Files Changed' page

File: crates/ironrdp-rdpdr-native/src/nix/mod.rs (hunk 5 above)
Comment (zmrush) [reply crates/ironrdp-rdpdr-native/src/nix/mod.rs]: ok, i added this explation in the newest commit. But when i change my code from 'as u32' to 'u32::try_from', it emit 'useless_conversion' warning, so i use '#[allow(clippy::useless_conversion)]' in the simple.rs and i add explanation.

**Comment by CBenoit** [comment]
>  but i do not know how to 'Also, please push new commits when applying review suggestions'. i only reply u on this 'Conversation' page and 'Files Changed' page

My suggestion was to create a new commit and push it normally instead of, e.g., amending it and force-push it. That way I can check the new commits instead, it’s much much easier to review.

**Comment by zmrush** [comment]
> > but i do not know how to 'Also, please push new commits when applying review suggestions'. i only reply u on this 'Conversation' page and 'Files Changed' page
> 
> My suggestion was to create a new commit and push it normally instead of, e.g., amending the initial one and force-pushing it each time. That way I can check the new commits instead, it’s much much easier to review.
> 
> When force-pushing, it appears like this in GitHub UI: image
> 
> I can squash all the commits into a single commit when merging, so it’s fine to push incremental review commits.

ok, i am sorry. i got it.

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 15]
```diff
@@ -0,0 +1,790 @@
...
+    pub file_id: u32,
+    pub file_base: String,
+    pub file_map: std::collections::HashMap<u32, std::fs::File>,
+    pub file_path_map: std::collections::HashMap<u32, String>,
+    pub file_dir_map: std::collections::HashMap<u32, DirBox>,
+}
+
+#[derive(Debug)]
+pub struct DirBox {
+    pub dir: *mut DIR,
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:27]: naming:
```suggestion
    pub ptr: *mut DIR,
```

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 16]
```diff
@@ -0,0 +1,790 @@
...
+}
+// SAFETY: it is safe to send *DIR to a different thread, and it is only accessed by one thread at one time.
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:29]: issue: It’s easy to break this assumption with the current API, the `dir` field is `pub` and the user can do whatever he wants with it:

```rust
let dir1 = DirBox { /* … */ }
let dir2 = DirBox { ptr: dir1.ptr) }
```

At this point, if you send `dir2` to a different thread, we are breaking the thread-safety. I suggest to not leave the field public and encapsulate it better.

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 17]
```diff
@@ -0,0 +1,790 @@
...
+                            if dir_name_in_rs == "." || dir_name_in_rs == ".." {
+                                continue;
+                            }
+                            parent.push_str(&dir_name_in_rs);
+                            find_file_name = Some(parent);
+                            break;
+                        }
+                        backend
+                            .file_dir_map
+                            .insert(req_inner.device_io_request.file_id, DirBox { dir: dirp });
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:585]: question: What happens when we keep around a DirBox with a null pointer? Is it intended? Shouldn’t we use an `Option<DirBox>` instead? Do we need to keep track of this at all when it failed?

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 18]
```diff
@@ -0,0 +1,790 @@
...
+                    let len = query_path.len();
+                    // path ends with *, so its len > 0
+                    #[allow(clippy::arithmetic_side_effects)]
+                    parent.push_str(&query_path[0..len - 1]);
+                    let dirp = if let Ok(real_parent_c) = CString::new(parent.clone()) {
+                        // SAFETY: it is safe to call this on *nix platform
+                        unsafe { opendir(real_parent_c.as_ptr()) }
+                    } else {
+                        std::ptr::null::<DIR>().cast_mut()
+                    };
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:563]: suggestion: I suggest to encapsulate this logic in `BoxDir`, in a `new` function.

```rust
impl BoxDir {
    fn new(path: &Path) -> Result<Self> {
        /* … */    
    }
}
```

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 19]
```diff
@@ -0,0 +1,790 @@
...
+                            // SAFETY: it is safe to call readdir on *nix platform
+                            let first = unsafe { readdir(dirp) };
+                            if first.is_null() {
+                                break;
+                            }
+                            // SAFETY: it is safe to call this for rdp
+                            let dir_name = unsafe { &(*first).d_name };
+                            // SAFETY: it is safe to call this for rdp
+                            let dir_name_in_rs =
+                                unsafe { CStr::from_ptr(dir_name as *const i8).to_string_lossy().into_owned() };
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:575]: question: What do you mean by "for rdp"? This seems unrelated to the RDP protocol.

suggestion: In this case, it’s a FFI structure and we can only assume that the library is well-behaved.
```suggestion
                            // SAFETY: When a non-null value is returned, the pointed struct should be properly initialized by the library.
                            let dir_name = unsafe { &(*first).d_name };

                            // SAFETY: DIR::name field contains the null terminated filename.
                            // etc etc
                            let dir_name_in_rs =
                                unsafe { CStr::from_ptr(dir_name as *const i8).to_string_lossy().into_owned() };
```

Also check the safety section of `CStr::from_ptr`: https://doc.rust-lang.org/std/ffi/struct.CStr.html#safety

suggestion: You can also likely encapsulate this in `BoxDir` directly.

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 20]
```diff
@@ -0,0 +1,790 @@
...
+                make_query_dir_resp(
+                    find_file_name,
+                    req_inner.device_io_request,
+                    req_inner.file_info_class_lvl,
+                    true,
+                )
+            } else {
+                if let Some(dir_box) = backend.file_dir_map.get_mut(&req_inner.device_io_request.file_id) {
+                    // SAFETY: it is safe to call readdir in multiple threads
+                    let next = unsafe { readdir(dir_box.dir) };
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:602]: question: You probably didn’t mean that. `readdir` is [not thread-safe](https://man7.org/linux/man-pages/man3/readdir.3.html).
image

But you didn’t implement `Sync` on `DirBox`, so we’re not supposed to be sharing references across several threads.

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 21]
```diff
@@ -0,0 +1,790 @@
...
+            let res = RdpdrPdu::DeviceCreateResponse(DeviceCreateResponse {
+                device_io_reply: io_response,
+                file_id,
+                information: Information::empty(),
+            });
+            Ok(vec![SvcMessage::from(res)])
+        }
+    }
+}
+pub(crate) fn process_dependent_file(
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:780]: style: Use some newlines to separate functions/types definitions.
```suggestion
}

pub(crate) fn process_dependent_file(
```

**Review by CBenoit — COMMENTED** [review]
`SAFETY` comments should be based on the documentation of the C functions. You can find that at different places, like:

- https://www.gnu.org/software/libc/manual/html_node/Reading_002fClosing-Directory.html
- https://man7.org/linux/man-pages/man3/readdir.3.html

You’ll find important information such as thread-safety, preconditions, postconditions, etc, etc. When there is nothing in particular which needs our attention, you can just justify with something like `FFI call without precondition`. When accessing fields of a struct returned by a C function, you can often use the postconditions of the function found in the doc and assume that the function is well-behaved.

File: crates/ironrdp-rdpdr-native/Cargo.toml [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+doctest = false
+test = false
+
+
+[target.'cfg(any(target_os = "macos", target_os = "linux"))'.dependencies]
+ironrdp-pdu.workspace = true
+ironrdp-svc.workspace = true
+ironrdp-rdpdr.workspace = true
+nix = { version = "0.29", features = ["fs"] }
+libc = "0.2"
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/Cargo.toml:24]: suggestion: You don’t need to specify this dependency at all since it’s re-exported by `nix`.

```suggestion
```

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 23]
```diff
@@ -0,0 +1,790 @@
...
+pub struct SimpleRdpdrBackend {
+    pub file_id: u32,
+    pub file_base: String,
+    pub file_map: std::collections::HashMap<u32, std::fs::File>,
+    pub file_path_map: std::collections::HashMap<u32, String>,
+    pub file_dir_map: std::collections::HashMap<u32, DirBox>,
+}
+
+#[derive(Debug)]
+pub struct DirBox {
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:26]: question: Maybe we can replace most of the unsafe code by using [`nix::dir::Dir`](https://docs.rs/nix/latest/nix/dir/struct.Dir.html). Did you consider it?

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 24]
```diff
@@ -0,0 +1,790 @@
...
+    req_inner: DeviceCloseRequest,
+) -> PduResult<Vec<SvcMessage>> {
+    backend.file_map.remove(&req_inner.device_io_request.file_id);
+    backend.file_path_map.remove(&req_inner.device_io_request.file_id);
+    if let Some(dir_box) = backend.file_dir_map.remove(&req_inner.device_io_request.file_id) {
+        // SAFETY: it is safe to close dir when we close device
+        unsafe {
+            closedir(dir_box.dir);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (zmrush) [reply crates/ironrdp-rdpdr-native/src/nix/simple.rs:179]: yes, add a drop function is on my schedule, but i am working on windows build and forgot it.  and as u suggest, nix::dir::Dir is better,  i modify the code

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 25]
```diff
@@ -0,0 +1,778 @@
...
+                        let res = RdpdrPdu::ClientDriveSetInformationResponse(ClientDriveSetInformationResponse::new(
+                            &req_inner,
+                            NtStatus::UNSUCCESSFUL,
+                        )?);
+                        return Ok(vec![SvcMessage::from(res)]);
+                    }
+                }
+                FileInformationClass::EndOfFile(info) => {
+                    let mut set_end_res = -1;
+                    // SAFETY: it is safe to call on *nix platform
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:414]: question: This is not really documenting why the unsafe operation is indeed safe. Per the documentation, it would be something like that:

```suggestion
                    // SAFETY: `file` is a regular file that we opened with write access.
```

However, I see `file_map` may contain files with `read` permission only. (See line 699.)
Can you explain how we are never calling `ftruncate` on these?

**Review by CBenoit — COMMENTED** [review]
Thank you! We can merge the PR once the last two comments are addressed! :slightly_smiling_face: 

Also: why `SimpleRdpdrBackend`? Do you think about implementing several backends for various use cases?

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs (hunk 25 above)
Comment (zmrush) [reply crates/ironrdp-rdpdr-native/src/nix/simple.rs:414]: ok, i got it

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 26]
```diff
@@ -0,0 +1,778 @@
...
+                    unsafe {
+                        if let Some(file) = backend.file_map.get(&req_inner.device_io_request.file_id) {
+                            set_end_res = nix::libc::ftruncate(file.as_raw_fd(), info.end_of_file);
+                        }
+                    }
+                    if set_end_res < 0 {
+                        warn!("set file end error, code is:{}", set_end_res);
```
[the remark this answers is not in the document]
Comment (zmrush) [reply crates/ironrdp-rdpdr-native/src/nix/simple.rs]: ok

**Comment by zmrush** [comment]
> Thank you! We can merge the PR once the last two comments are addressed! 🙂
> 
> Also: why `SimpleRdpdrBackend`? Do you think about implementing several backends for various use cases?

because, i has not implements all the features. In addition, i can implements more features. for example 'query directory information', the file information is return one by one, which is not efficient.

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 27]
```diff
@@ -0,0 +1,780 @@
...
+                    let mut set_end_res = -1;
+                    // SAFETY: it is safe to call on *nix platform
+                    // SAFETY: `file` is a regular file that we opened with write access.
+                    unsafe {
+                        if let Some(file) = backend.file_map.get(&req_inner.device_io_request.file_id) {
+                            set_end_res = nix::libc::ftruncate(file.as_raw_fd(), info.end_of_file);
+                        }
+                    }
+                    if set_end_res < 0 {
+                        let error = nix::errno::Errno::last(); // However, only call this if `ftruncate` was called.
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:422]: issue: The comment wasn’t intended to be actually included. You should modify the code to not call `Errono::last()` unless we actually called `ftruncate`.

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 28]
```diff
@@ -0,0 +1,780 @@
...
+                            &req_inner,
+                            NtStatus::UNSUCCESSFUL,
+                        )?);
+                        return Ok(vec![SvcMessage::from(res)]);
+                    }
+                }
+                FileInformationClass::EndOfFile(info) => {
+                    let mut set_end_res = -1;
+                    // SAFETY: it is safe to call on *nix platform
+                    // SAFETY: `file` is a regular file that we opened with write access.
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:415]: issue: The first comment is not about safety so you can remove it. Also you forgot to explain why we know that the file is actually opened with write access. I see that `file_map` may contain files with read permission only. (See line 699.) Can you explain how we are never calling ftruncate on these?

```suggestion
                    // SAFETY: `file` is a regular file that we opened with write access.
                    // We know that `file` is opened with write access because… [explain why]
```

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs [hunk 29]
```diff
@@ -0,0 +1,780 @@
...
+                    unsafe {
+                        if let Some(file) = backend.file_map.get(&req_inner.device_io_request.file_id) {
+                            set_end_res = nix::libc::ftruncate(file.as_raw_fd(), info.end_of_file);
+                        }
+                    }
```
Comment (CBenoit) [inline crates/ironrdp-rdpdr-native/src/nix/simple.rs:420]: suggestion: Keep unsafe blocks as small as possible.
```suggestion
                    if let Some(file) = backend.file_map.get(&req_inner.device_io_request.file_id) {
                         unsafe {
                            set_end_res = nix::libc::ftruncate(file.as_raw_fd(), info.end_of_file);
                        }
                    }
```

**Comment by CBenoit** [comment]
> > Thank you! We can merge the PR once the last two comments are addressed! 🙂
> > Also: why `SimpleRdpdrBackend`? Do you think about implementing several backends for various use cases?
> 
> because, i has not implements all the features. In addition, i can implements more features. for example 'query directory information', the file information is return one by one, which is not efficient.

Even if not all features are implemented now, we could implement them in the future. I’m fine with dropping the `Simple` part given that we’ll likely just extend the current code, instead of creating a separate backend. Or do you think we should keep a "simple" backend that is not supporting specific features?

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs (hunk 28 above)
Comment (zmrush) [reply crates/ironrdp-rdpdr-native/src/nix/simple.rs]: i am sorry for not understanding your suggestion. and i modify it

File: crates/ironrdp-rdpdr-native/src/nix/simple.rs (hunk 29 above)
Comment (zmrush) [reply crates/ironrdp-rdpdr-native/src/nix/simple.rs]: ok,i got it and modify it

**Comment by zmrush** [comment]
> > > Thank you! We can merge the PR once the last two comments are addressed! 🙂
> > > Also: why `SimpleRdpdrBackend`? Do you think about implementing several backends for various use cases?
> > 
> > 
> > because, i has not implements all the features. In addition, i can implements more features. for example 'query directory information', the file information is return one by one, which is not efficient.
> 
> Even if not all features are implemented now, we could implement them in the future. I’m fine with dropping the `Simple` part given that we’ll likely just extend the current code, instead of creating a separate backend. Or do you think we should keep a "simple" backend that is not supporting specific features?

ok,i also think it is fine, i modify the name.

**Review by CBenoit — APPROVED** [review]
Thank you for this work!
